rust/core-lib/src/view.rs

```diff
@@ -517,6 +590,175 @@ impl View {
+            let num_steps = if progress { 32 } else { usize::max_value() };
+
+            loop {
+                match find_progress(&mut cursor, self.case_matching, &search_string, num_steps) {
+                    FindResult::Found(pos) => {
+                        let region = SelRegion {
+                            start: pos,
+                            end: pos + len,
+                            horiz: None,
+                            affinity: Affinity::default(),
+                        };
+                        occurrences.add_region(region);
+                    }
+                    FindResult::NotFound => break,
+                    FindResult::TryAgain => {
+                        searched_until = cursor.pos();
+                        break;
+                    }
+                }
+            }
+        }
+        self.occurrences = Some(occurrences);
+        self.valid_search.union_one_range(start, searched_until);
+        self.hls_dirty = true;
+    }
+
+    /// Select the next occurrence relative to the last cursor. `reverse` determines whether the
+    /// next occurrence before (`true`) or after (`false`) the last cursor is selected. `wrapped`
+    /// indicates a search for the next occurrence past the end of the file. `allow_stay` determines
+    /// whether
+    pub fn select_next_occurrence(&mut self, text: &Rope, reverse: bool, wrapped: bool) -> Option<usize> {
+        if self.search_string.is_none() {
+            return None;
+        }
+
+        let sel = match self.sel_regions().iter().last() {
+            Some(sel) => (sel.min(), sel.max()),
+            None => return None
+        };
+        
+        let (from, to) = if reverse != wrapped { (0, sel.0) } else { (sel.1, text.len()) };
+        let mut next_occurrence;
+
+        loop {
+            next_occurrence = self.occurrences.as_ref().and_then(|occurrences| {
+                if occurrences.len() == 0 {
+                    return None;
+                }
+                if wrapped {
+                    if reverse {
+                        Some(occurrences[occurrences.len() - 1].clone())
+                    } else {
+                        Some(occurrences[0].clone())
+                    }
+                } else {
+                    let mut ix = occurrences.search(sel.0);
+                    if reverse {
+                        if ix == 0 {
+                            None
+                        } else {
+                            Some(occurrences[ix-1].clone())
+                        }
+                    } else {
+                        if !wrapped && ix < occurrences.len() && occurrences[ix].end == sel.1 {
+                            ix += 1;
+                        }
+                        if ix >= occurrences.len() {
+                            None
+                        } else {
+                            Some(occurrences[ix].clone())
+                        }
+                    }
+                }
```

Comment:
- [ ] I've to admit that this block (above) is really ugly (in terms of readability and understandability). I may try to come up with a simplified logic or add some meaningful code comments ...